src/convert.ts

```diff
@@ -134,3 +136,212 @@ export function convertCatalog(
   }
   return result;
 }
+
+export type Share = { version: string; initSources: any[] };
+export type ShareResult = {
+  result: Share | null;
+  messages?: Message[];
+  converted?: boolean;
+};
+export type Story = {
+  title: string;
+  text: string;
+  id: string;
+  shareData: Share;
+};
+export function convertShare(json: unknown): ShareResult {
+  if (!is.plainObject(json)) {
+    return {
+      result: null,
+      messages: [inputNotPlainObject()],
+    };
+  }
+
+  // If version 8 return
+  if (
+    "version" in json &&
+    typeof json.version === "string" &&
+    json.version.startsWith("8")
+  ) {
+    return { result: json as Share, converted: false };
+  }
+
+  if (!Array.isArray(json.initSources)) {
+    return {
+      result: null,
+      messages: [
+        missingRequiredProp(
+          ModelType.Share,
+          "initSources",
+          undefined,
+          "Init sources"
+        ),
+      ],
+    };
+  }
+
+  const messages: Message[] = [];
+
+  const initializationUrls: string[] = [];
+
+  // Crunch v7 initSources together + extract initializationUrls
+  const v7InitSource = json.initSources.reduce<any>((sources, current) => {
+    if (typeof current === "string") {
+      initializationUrls.push(current);
+      return sources;
+    }
+
+    return merge(sources, current);
```

Comment:
Sorry, I didn't make this clear. This should use the algorithm in `mergeRecursive.js`. The file is checked into this repo. It's very specific to Terria v7 catalogues as there are some odd things about how members are merged. This merging is heavily relied upon in the nationalmap catalogue (that's why https://nationalmap.dev.saas.terria.io/ has such a weird catalogue).

Reply:
To fix the nationalmap problem we'll need to use `mergeRecursive` on any input to the converter. It'd be nice if we only ran it on each sub-tree of the catalogue once (this could be implemented with a parameter `mergeItems` or similar that's default `true` when you pass a member/list of members to one of the convert functions and then any recursive calls use `false`), but it doesn't really matter if it runs multiple times so we don't have to add that complexity.

Reply:
Hey @steve9164 , sorry I'm not sure where `mergeRecursive` is. I can't find a `mergeRecursive.js` file, and it doesn't seem to be in any libs. Where do I find it?

Reply:
Fair call. I thought I committed some catalog build tools (used in droughtmap, investormap and others) to catalog-converter, but I didn't commit them. Here's mergeRecursive:
```js
"use strict";

function mergeRecursive(catalogMembers) {
  // Recursively merge a list catalog members to a list of catalog members that is equivalent when loaded with TerriaJS
  const members = [];
  catalogMembers.forEach(m => {
    // Find a matching member and merge properties, or add the member m
    // Assuming that a null id doesn't count as an id
    let matchIndex = -1;

    // ID matching commented out because it doesn't match TerriaJS behaviour
    // if (m.id != null) { // Check not null or undefined
    //   matchIndex = members.findIndex(m2 => m2.id === m.id);
    // } else {
    matchIndex = members.findIndex(
      m2 => /*m2.id == null &&*/ m2.name === m.name
    );
    // }

    const existingM = matchIndex >= 0 ? members[matchIndex] : {};
    // Merge m and existingM
    // For a catalog group, concat items and merge properties
    // For anything else just merge properties
    const updatedM = Object.assign({}, existingM, m);
    if (existingM.items != null || m.items != null) {
      updatedM.items = mergeRecursive(
        [].concat(existingM.items || [], m.items || [])
      );
    }
    if (matchIndex >= 0) {
      members[matchIndex] = updatedM;
    } else {
      members.push(updatedM);
    }
  });
  return members;
}

module.exports = mergeRecursive;
```

Reply:
We decided that this is not needed for sharelinks